### PR TITLE
feat: add type inference for `InvokeMethod2` with parametered methods

### PR DESCRIPTION
**Pull request by chanattan**
Work continuation of #7332.

### CODE REVIEW

**Comment by chanattan** [comment]
@magnus-madsen

**Comment by magnus-madsen** [comment]
Lets start adding a battery of tests at the same time we extend the expressive power.

**Comment by chanattan** [comment]
Ok I'll do it in a separate PR.

**Comment by magnus-madsen** [comment]
I prefer to do it in this pr. From now on, we extend the compiler and we add the tests at the same time. That's how we usually do it. 👍

File: main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction.scala [hunk 1]
```diff
@@ -134,20 +134,31 @@ object TypeReduction {
...
+   * Helper method to retrieve a method given its name and parameter types.
+   * Returns a JavaResolutionResult either containing the Java method or a MethodNotFound object.
+   */
+  private def retrieveMethod(clazz: Class[_], method: String, ts: List[Type], loc: SourceLocation)(implicit flix: Flix): JavaResolutionResult = {
+    // NB: this considers also static methods
+    val candidateMethods = clazz.getMethods
+      .filter(m => m.getName == method)
+      .filter(m => m.getParameterCount == ts.length)
+      .filter(m =>
+        (m.getParameterTypes zip ts).foldLeft(true){ case (acc, (clazz, tpe)) => acc && (Type.getFlixType(clazz) == tpe) } // Parameter types correspondance
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction.scala:151]: Later we will need to consider subtyping here.

File: main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction.scala [hunk 2]
```diff
@@ -134,20 +134,31 @@ object TypeReduction {
...
+      )
+
+    candidateMethods.length match {
+      case 1 =>
+        val tpe = Type.Cst(TypeConstructor.JvmMethod(candidateMethods.head), loc)
+        JavaResolutionResult.Resolved(tpe)
+      case _ => JavaResolutionResult.MethodNotFound // For now, method not found either if there is an ambiguity or no method found
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction.scala:158]: Yes, ambiguity later.

**Review by magnus-madsen — APPROVED** [review]

**Comment by magnus-madsen** [comment]
@chanattan Will merge this PR when tests complete. Then you can iterate on the next feature.

**Comment by chanattan** [comment]
Actually I wanted to commit overloading tests too.
However I have some trouble when using string split to be equal to some Array in Flix. @magnus-madsen

**Comment by magnus-madsen** [comment]
We cannot deal with arrays yet. We leave that for a bit later. Its complicated :)

**Pull request merged**